### PR TITLE
fix: transfer error when reset data in some cases

### PR DESCRIPTION
**Pull request by BoyYangzai**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
- fix https://github.com/ant-design/ant-design/issues/42782

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   fix: transfer error when reset data in some cases    |
| 🇨🇳 Chinese |      修复一些 cases 下重置数据出现 seletcedKeys 错误的情况     |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at fe08ed8</samp>

Fixed a bug in the `transfer` component that caused an error when the data source and the target keys were reset. Added a test case to verify the fix and improved the logic of the `useEffect` hook in `components/transfer/index.tsx`.

### <samp>🤖 Generated by Copilot at fe08ed8</samp>

*  Fix error when reset data in transfer component ([link](https://github.com/ant-design/ant-design/pull/42785/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL162-R175), [link](https://github.com/ant-design/ant-design/pull/42785/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296L2-R8), [link](https://github.com/ant-design/ant-design/pull/42785/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296R626-R693))
  - Add imports of `DefaultRecordType` and `Button` in test file `components/transfer/__tests__/index.test.tsx` ([link](https://github.com/ant-design/ant-design/pull/42785/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296L2-R8))
  - Modify `useEffect` hook in `components/transfer/index.tsx` to handle no `props.selectedKeys` case ([link](https://github.com/ant-design/ant-design/pull/42785/files?diff=unified&w=0#diff-16f601b16a2f3b1d448af36a6b569c1d47bf9e2c351318a3b986c42fcbc5b7fdL162-R175))
  - Add test case in `components/transfer/__tests__/index.test.tsx` to check no error when data source and target keys are reset by a button in the footer ([link](https://github.com/ant-design/ant-design/pull/42785/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296R626-R693))

### CODE REVIEW

**Comment by afc163** [comment]
看下这里的报错提示：https://github.com/ant-design/ant-design/pull/42785/files

File: components/transfer/index.tsx [hunk 1]
```diff
@@ -159,6 +159,19 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
     if (props.selectedKeys) {
       setSourceSelectedKeys(() => selectedKeys.filter((key) => !targetKeys.includes(key)));
       setTargetSelectedKeys(() => selectedKeys.filter((key) => targetKeys.includes(key)));
+      return;
+    }
+
+    // This is the calculation scheme when there are no props.selectedKeys
+    if (props.selectedKeys) {
```
Comment (zombieJ) [inline components/transfer/index.tsx:166]: 上面有一个 `props.selectedKeys` + `return`，这段代码应该跑不到的

File: components/transfer/index.tsx [hunk 2]
```diff
@@ -155,12 +155,37 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
+      setSourceSelectedKeys(propSourceKeys.filter((key) => sourceSelectedKeys.includes(key)));
+    }
+
+    if (props.targetKeys) {
+      setTargetSelectedKeys(() =>
+        props.targetKeys!.filter((key) => targetSelectedKeys.includes(key)),
+      );
     }
-  }, [props.selectedKeys, props.targetKeys]);
+  }, [props.selectedKeys, props.targetKeys, dataSource]);
```
Comment (zombieJ) [inline components/transfer/index.tsx:188]: 这里没看到用到过 dataSource，依赖是 propSourceKeys？

File: components/transfer/index.tsx [hunk 3]
```diff
@@ -155,12 +155,37 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
+        })
+        .filter(Boolean) as string[],
+    [dataSource],
+  );
+
   useEffect(() => {
     if (props.selectedKeys) {
       setSourceSelectedKeys(() => selectedKeys.filter((key) => !targetKeys.includes(key)));
       setTargetSelectedKeys(() => selectedKeys.filter((key) => targetKeys.includes(key)));
+      return;
```
Comment (zombieJ) [inline components/transfer/index.tsx:175]: 不要用 return，容易埋坑，用 elseif

**Comment by zombieJ** [comment]
感觉这个实现不太好，太多的  `setSourceSelectedKeys`，最好是直接给 sourceSelectedKeys 做个 memo 过滤，这样统一收口

File: components/transfer/index.tsx (hunk 2 above)
Comment (afc163) [reply components/transfer/index.tsx:188]: 把 keys join 出来，变成 string。

File: components/transfer/index.tsx [hunk 4]
```diff
@@ -165,27 +165,54 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
+    () =>
+      getMergedKeys(
+        selectedKeys,
+        targetKeys,
+        targetSelectedKeys,
+        props.targetKeys,
+        props.selectedKeys,
+        false,
+      ),
+    [selectedKeys, targetKeys, targetSelectedKeys, props.targetKeys, props.selectedKeys],
```
Comment (afc163) [inline components/transfer/index.tsx:214]: 都改成 `XxxKeys.join(',')`

File: components/transfer/index.tsx [hunk 5]
```diff
@@ -155,12 +155,64 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
-      setSourceSelectedKeys(() => selectedKeys.filter((key) => !targetKeys.includes(key)));
-      setTargetSelectedKeys(() => selectedKeys.filter((key) => targetKeys.includes(key)));
+  const propSourceKeys = useMemo(
+    () =>
+      dataSource
+        .map((item) => {
+          if (!props.targetKeys?.includes(item.key!)) {
+            return item.key;
+          }
+          return false;
```
Comment (li-jia-nan) [inline components/transfer/index.tsx:165]: 这里 `return false` 有点奇怪，虽然是为了后面 `filter(Boolean)`，但是感觉可以先用 filter 把需要的 item 取到，最后再用 `map(item => item.key)`，你觉得呢？

File: components/transfer/index.tsx [hunk 6]
```diff
@@ -148,19 +148,49 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
   const [targetSelectedKeys, setTargetSelectedKeys] = useState<string[]>(() =>
-    selectedKeys.filter((key) => targetKeys.includes(key)),
+    (selectedKeys ?? []).filter((key) => targetKeys.includes(key)),
+  );
+
+  const propSourceKeys = useMemo(
+    () =>
+      dataSource
+        .filter((item) => !props.targetKeys?.includes(item.key!))
+        .map((item) => item.key) as string[],
```
Comment (li-jia-nan) [inline components/transfer/index.tsx:162]:
```suggestion
        .map((item) => item.key!),
```

File: components/transfer/index.tsx [hunk 7]
```diff
@@ -148,19 +148,49 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
-    selectedKeys.filter((key) => !targetKeys.includes(key)),
+    (selectedKeys ?? []).filter((key) => !targetKeys.includes(key)),
   );
 
   const [targetSelectedKeys, setTargetSelectedKeys] = useState<string[]>(() =>
-    selectedKeys.filter((key) => targetKeys.includes(key)),
+    (selectedKeys ?? []).filter((key) => targetKeys.includes(key)),
+  );
+
+  const propSourceKeys = useMemo(
```
Comment (li-jia-nan) [inline components/transfer/index.tsx:158]:
```suggestion
  const propSourceKeys = useMemo<string[]>(
```

File: components/transfer/index.tsx (hunk 4 above)
Comment (li-jia-nan) [reply components/transfer/index.tsx]: 建议用 `JSON.stringify`，稍微复杂一点的数据也能用

Comment (BoyYangzai) [reply components/transfer/index.tsx]: 这里全是 string[] 不用太复杂

File: components/transfer/index.tsx [hunk 8]
```diff
@@ -148,19 +148,52 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
+    if (propTypeKeys) {
+      return propTypeKeys.filter((key) => typeSelectedKeys.includes(key));
+    }
+    return keysWithSelectedOrNoTarget;
+  };
+
+  const mergedSourceSelectedKeys = useMemo(
+    () => getMergedKeys(targetKeys, sourceSelectedKeys, propSourceKeys, selectedKeys, true),
+    [
+      selectedKeys?.join(''),
```
Comment (afc163) [inline components/transfer/index.tsx:186]:
```suggestion
      selectedKeys?.join(','),
```

加个分隔符，否则 `[12, 3]` 和 `[1, 23]` 分不开。

Comment (BoyYangzai) [reply components/transfer/index.tsx]: done

File: components/transfer/index.tsx [hunk 9]
```diff
@@ -148,19 +148,49 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
   const [wrapSSR, hashId] = useStyle(prefixCls);
 
   const [sourceSelectedKeys, setSourceSelectedKeys] = useState<string[]>(() =>
-    selectedKeys.filter((key) => !targetKeys.includes(key)),
+    (selectedKeys ?? []).filter((key) => !targetKeys.includes(key)),
```
Comment (zombieJ) [inline components/transfer/index.tsx:151]: 这个改动似乎没有必要，`??` 会加大 bundle size

File: components/transfer/index.tsx [hunk 10]
```diff
@@ -148,19 +148,49 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
...
+    () =>
+      dataSource.filter((item) => !props.targetKeys?.includes(item.key!)).map((item) => item.key!),
+    [dataSource, props.targetKeys],
   );
 
-  useEffect(() => {
-    if (props.selectedKeys) {
-      setSourceSelectedKeys(() => selectedKeys.filter((key) => !targetKeys.includes(key)));
-      setTargetSelectedKeys(() => selectedKeys.filter((key) => targetKeys.includes(key)));
+  const getMergedKeys = (
```
Comment (zombieJ) [inline components/transfer/index.tsx:164]: 这个 function 的参数命名看的有点懵逼，最好在语义化一些

**Review by li-jia-nan — APPROVED** [review]

**Pull request merged**